content/blog/2022-04-07-compiling-black-with-mypyc-part-2.md

```diff
@@ -0,0 +1,537 @@
+
+-    def addtoken(self, type: int, value: Optional[Text], context: Context) -> bool:
++    def addtoken(self, type: int, value: Text, context: Context) -> bool:
+         """Add a token; return True iff this is the end of the program."""
+         # Map from token to label
+         ilabel = self.classify(type, value, context)
+@@ -185,11 +185,10 @@ class Parser(object):
+                     # No success finding a transition
+                     raise ParseError("bad input", type, value, context)
+
+-    def classify(self, type: int, value: Optional[Text], context: Context) -> int:
++    def classify(self, type: int, value: Text, context: Context) -> int:
+         """Turn a token into a label.  (Internal)"""
+         if type == token.NAME:
+             # Keep a listing of all used names
+-            assert value is not None
+             self.used_names.add(value)
+             # Check for reserved words
+             ilabel = self.grammar.keywords.get(value)
+@@ -201,12 +200,10 @@ class Parser(object):
+         return ilabel
+
+     def shift(
+-        self, type: int, value: Optional[Text], newstate: int, context: Context
++        self, type: int, value: Text, newstate: int, context: Context
+     ) -> None:
+         """Shift a token.  (Internal)"""
+         dfa, state, node = self.stack[-1]
+-        assert value is not None
+-        assert context is not None
+         rawnode: RawNode = (type, value, context, None)
+         newnode = self.convert(self.grammar, rawnode)
+         if newnode is not None:
+```
+
+This also has the neat side-effect of allowing me to remove some asserts.
+
+Tightening up type annotations involving [`typing.Any`][typing-any] can be particularly
+worthwhile as `Any` forces mypyc to fallback to generic C code that can handle any kind of
+object. I was only able to find one spot I could make this change, but it's better than
+nothing!
+
+```diff
+@@ -54,14 +56,14 @@ class Driver(object):
+         self.logger = logger
+         self.convert = convert
+
+-    def parse_tokens(self, tokens: Iterable[Any], debug: bool = False) -> NL:
++    def parse_tokens(self, tokens: Iterable[GoodTokenInfo], debug: bool = False) -> NL:
+         """Parse a series of tokens and return the syntax tree."""
+         # XXX Move the prefix computation into a wrapper around tokenize.
+         p = parse.Parser(self.grammar, self.convert)
+         p.setup()
+         lineno = 1
+         column = 0
+```
+
+#### Marking everything Final
+
+Avoiding calculating the same value over and over again is one of the most common
+optimizations out there, and it's for good reason, it's usually very easy to fix. BUT,
+with mypyc we can take this further using [`typing.Final`][typing-final]. Final variables
+can often be injected at lookup sites at compile time, skipping the lookups at runtime!
+
+Let me show an example, let's take this code and see what adding a single `Final` does.
+
+```python3
+from typing import Final
```

Comment:
Consider taking this line out, since it confused me the first time I saw it (since it's not being used) until later when you add `Final`